### DISCUSSION TITLE
Deploy to Heroku and environment variables

### DISCUSSION THREAD

**Discussion by yoxud** [saleor/saleor#7133] [Deployment] [answered]
Trying to deploy saleor to Heroku according to the Saleor Website doing 

`heroku git:remote -a '<saleor-app name>'
heroku buildpacks:set heroku/nodejs
heroku buildpacks:add heroku/python
heroku addons:create heroku-postgresql:hobby-dev
heroku addons:create heroku-redis:hobby-dev
heroku addons:create sendgrid:starter
heroku config:set ALLOWED_HOSTS='<your hosts here>'
heroku config:set NODE_MODULES_CACHE=false
heroku config:set NPM_CONFIG_PRODUCTION=false
heroku config:set SECRET_KEY='<your secret key here>
git push heroku master
`

The push fails with the following error
`remote: -----> $ python manage.py collectstatic --noinput
remote:        Traceback (most recent call last):
remote:          File "/app/.heroku/python/lib/python3.9/site-packages/django/core/management/__init__.py", line 224, in fetch_command
remote:            app_name = commands[subcommand]
remote:        KeyError: 'collectstatic'
remote:        During handling of the above exception, another exception occurred:
remote:        Traceback (most recent call last):
remote:          File "/tmp/build_efd670fc/manage.py", line 10, in <module>
remote:            execute_from_command_line(sys.argv)
remote:          File "/app/.heroku/python/lib/python3.9/site-packages/django/core/management/__init__.py", line 401, in execute_from_command_line
remote:            utility.execute()
remote:          File "/app/.heroku/python/lib/python3.9/site-packages/django/core/management/__init__.py", line 395, in execute
remote:            self.fetch_command(subcommand).run_from_argv(self.argv)
remote:          File "/app/.heroku/python/lib/python3.9/site-packages/django/core/management/__init__.py", line 231, in fetch_command
remote:            settings.INSTALLED_APPS
remote:          File "/app/.heroku/python/lib/python3.9/site-packages/django/conf/__init__.py", line 82, in __getattr__
remote:            self._setup(name)
remote:          File "/app/.heroku/python/lib/python3.9/site-packages/django/conf/__init__.py", line 69, in _setup
remote:            self._wrapped = Settings(settings_module)
remote:          File "/app/.heroku/python/lib/python3.9/site-packages/django/conf/__init__.py", line 170, in __init__
remote:            mod = importlib.import_module(self.SETTINGS_MODULE)
remote:          File "/app/.heroku/python/lib/python3.9/importlib/__init__.py", line 127, in import_module
remote:            return _bootstrap._gcd_import(name[level:], package, level)
remote:          File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
remote:          File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
remote:          File "<frozen importlib._bootstrap>", line 972, in _find_and_load_unlocked
remote:          File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
remote:          File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
remote:          File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
remote:          File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
remote:          File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
remote:          File "<frozen importlib._bootstrap_external>", line 790, in exec_module
remote:          File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
remote:          File "/tmp/build_efd670fc/saleor/__init__.py", line 1, in <module>
remote:            from .celeryconf import app as celery_app
remote:          File "/tmp/build_efd670fc/saleor/celeryconf.py", line 6, in <module>
remote:            from .plugins import discover_plugins_modules
remote:        ModuleNotFoundError: No module named 'saleor.plugins'
remote:
remote:  !     Error while running '$ python manage.py collectstatic --noinput'.
remote:        See traceback above for details.
remote:
remote:        You may need to update application code to resolve this error.
remote:        Or, you can disable collectstatic for this application:
remote:
remote:           $ heroku config:set DISABLE_COLLECTSTATIC=1
remote:
remote:        https://devcenter.heroku.com/articles/django-assets
remote:  !     Push rejected, failed to compile Python app.`

I can only get it to push by executing as suggested

` $ heroku config:set DISABLE_COLLECTSTATIC=1`

But then it seems that the environment variables are somehow statically set at the first build time and I can not modify them subsequently through heroku. the heroku config shows all modifications but the app ignores them. A subsequent push does not modify the original environment variables. The only way I found was to delete the heroku app and reinstall everything... So my question is how can I deploy to Heroku and subsequently be able to use and modify the heroku environment dynamically?

**Comment by krzysztofwolski** [comment]
Moving deployment question to discussions section.

**Comment by karolkielecki** [comment]
@yoxud Hi there,

Some of our community members experience issues with Heroku deployment at the beginning. Something went wrong with your deployment--that's for sure, but it's not a Saleor issue. Remember, you need to create a new Heroku app first. Please find more information on Heroku deployment here: https://devcenter.heroku.com/categories/deployment

**Comment by yoxud** [comment, marked as answer]
@karolkielecki , These deployment errors are due to the fact that my working tree was missing files due to an incomplete checkout of the repo due to issue  #7135 . It is a bit insidious as the cloning process succeeds but there is only a warning for the checkout failure and since I don't know much django lingo,  It took me some time to realize I was missing files... Anyway this discussion can be closed.
